### ISSUE TITLE
home-manager extraCss option for more targets

### ISSUE THREAD

**Issue by jamesbaw1** [danth/stylix#481]
Currently, it's quite difficult to add CSS on top of the CSS generated by Stylix. The main issue is when it comes to overwriting CSS rules, because the user-defined CSS (for example programs.waybar.style in home.nix) is applied before the CSS generated by Stylix.

Perhaps we could add an extraCss option to more (hopefully eventually all) targets, as it is already a feature for GTK, and would be as simple (I think?) as appending the text in the user-defined extraCss to the CSS that is already being applied to the target. For example:

```
programs.waybar.style =
  # ...
  + (if config.stylix.targets.waybar.extraCss then stylix.targets.waybar.extraCss else ""); # I don't think this code works, but I hope It's somewhat illustrative
```

This could also act as a band-aid solution to #429 and maybe some other feature requests concerning CSS, as users could simply overwrite existing CSS rules.

The other option may be to somehow get the user-defined programs.waybar.style to apply after Stylix's version, but I'm not an expert at nix, and I don't know how feasible this is. It's also not as universal a solution.

**Comment by danth** [comment]
There are `mkBefore` and `mkAfter` functions in `lib`. These change the order in which option definitions are merged; for example the two definitions `"a"` and `mkBefore "b"` would be guaranteed to merge as `"ba"`.

If we add `mkBefore` to any CSS defined by Stylix, then we can guarantee that CSS added by the user will be applied later without needing to create another option.

**Comment by jamesbaw1** [comment]
So,
```
config = lib.mkIf (config.stylix.enable && config.stylix.targets.waybar.enable) {
    programs.waybar.style = # ...
};
```
would become
```
config = lib.mkIf (config.stylix.enable && config.stylix.targets.waybar.enable) lib.mkBefore {
    programs.waybar.style = # ...
};
```
?

I'm relatively new to the nix language, so sorry if I'm missing the obvious.

**Comment by danth** [comment]
I believe the `mkBefore` needs to go on the `programs.waybar.style` option specifically, so it should be `programs.waybar.style = lib.mkBefore ...`

**Comment by jamesbaw1** [comment]
Ah, okay. I'm still getting my setup in order, so it may take a while for me to submit a PR.

**Comment by awwpotato** [comment]
I feel like this is unnecessary duplication of code, the user can already add more css with `program.<name>.style = lib.mkAfter "";`. Adding this to stylix would just lead to unnecessary code duplication.

**Comment by Flameopathic** [comment]
> I feel like this is unnecessary duplication of code, the user can already add more css with `program.<name>.style = lib.mkAfter "";`.

it would be if all of the options were added, but it doesn't seem like it would be if we simply followed danth's guidance:

> If we add `mkBefore` to any CSS defined by Stylix, then we can guarantee that CSS added by the user will be applied later without needing to create another option.

**Comment by awwpotato** [comment]
> but it doesn't seem like it would be if we simply followed danth's guidance:
> 
> > If we add `mkBefore` to any CSS defined by Stylix, then we can guarantee that CSS added by the user will be applied later without needing to create another option.

we have already decided not to apply mkBefore to the css defined by stylix (#388, https://github.com/danth/stylix/pull/817#pullrequestreview-2579115983)

**Comment by Flameopathic** [comment]
are there any further objections to adding the `lib.mkAfter` solution to the documentation and closing this issue?

**Comment by trueNAHO** [comment]
> are there any further objections to adding the `lib.mkAfter` solution to the documentation and closing this issue?

No. Feel free to submit a PR. This should probably be added to `/docs/src/tricks.md`.

**Comment by Flameopathic** [comment]
after going through files to find a module to use for demonstrating the use of `lib.mkAfter`, i've actually noticed that isn't actually useful very often. there are many modules which use `home.file` to set their CSS, which #704 seems to imply cannot be just merged with whatever the user sets.

**Comment by Flameopathic** [comment]
> > there are many modules which use home.file to set their CSS, which [#704](https://github.com/danth/stylix/issues/704) seems to imply cannot be just merged with whatever the user sets.
> 
> I don't know what your talking about... just because the home-manager module uses home.file under the hood doesn't mean the styles can't be merged. Take wofi for example, the `style` option is declared as `lib.types.lines`, which will merge the final product into a string. Then that merged style string is put into `xdg.configFile` which [puts that into `home.file`](https://github.com/nix-community/home-manager/blob/master/modules/misc/xdg.nix#L154)

the issue comes when Stylix itself is setting a theme using `home.file`. see the `discord` module, for example. in that case, the only conceivable way for the user to override or add to the styling would be to define their own `home.file` option, but #704 seems to imply that the Stylix and user declarations would not be merged and would instead cause an error.

**Comment by Flameopathic** [comment]
> > the issue comes when Stylix itself is setting a theme using `home.file`. see the `discord` module, for example. in that case, the only conceivable way for the user to override or add to the styling would be to define their own `home.file` option, but [#704](https://github.com/danth/stylix/issues/704) seems to imply that the Stylix and user declarations would not be merged and would instead cause an error.
> 
> Oh I get what your saying. In the modules that use home.file adding `extraCss` would be reasonable.

if we plan on adding `extraCss` to some modules, i think that it would logically follow that we do it treewide for the sake of consistency.

**Comment by awwpotato** [comment]
> if we plan on adding extraCss to some modules, i think that it would logically follow that we do it treewide for the sake of consistency.

I strongly disagree. Duplicating home-manager options is a waste of time and introduces more complexity.

**Comment by Flameopathic** [comment]
here is a list of modules which use `home.file` either directly or indirectly through `xdg`:

- btop
- discord
- fcitx5
- forge
- gedit
- gnome
- gtk
- halloy
- helix
- kde
- qt
- wezterm
- zed
- zellij

it may be best that, if we plan to add `extraCss` options for all of them, we do it in one PR

**Comment by awwpotato** [comment]
> here is a list of modules which use `home.file` either directly or indirectly through `xdg`:
> 
>     * btop
> 
>     * discord
> 
>     * fcitx5
> 
>     * forge
> 
>     * gedit
> 
>     * gnome
> 
>     * gtk
> 
>     * halloy
> 
>     * helix
> 
>     * kde
> 
>     * qt
> 
>     * wezterm
> 
>     * zed
> 
>     * zellij
> 
> 
> it may be best that, if we plan to add `extraCss` options for all of them, we do it in one PR

Most of these use `source =`, not `text =` which makes it more complicated to add extra{Style/Css/Whatever} to them

**Comment by Flameopathic** [comment]
> > here is a list of modules which use `home.file` either directly or indirectly through `xdg`:
> > ```
> > * btop
> > 
> > * discord
> > 
> > * fcitx5
> > 
> > * forge
> > 
> > * gedit
> > 
> > * gnome
> > 
> > * gtk
> > 
> > * halloy
> > 
> > * helix
> > 
> > * kde
> > 
> > * qt
> > 
> > * wezterm
> > 
> > * zed
> > 
> > * zellij
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > it may be best that, if we plan to add `extraCss` options for all of them, we do it in one PR
> 
> Most of these use `source =`, not `text =` which makes it more complicated to add extra{Style/Css/Whatever} to them

isn't that all the more reason to implement an option for it? it's not that hard, either; see https://github.com/danth/stylix/blob/489833b201a84488c6b4371a261fdbcafa6abcb6/modules/gtk/hm.nix#L17-L20

**Comment by danth** [comment]
It may be more productive to contribute new options to Home Manager for the modules where we're currently using `home.file`, then we can use `lib.mkAfter` on them as we do for other modules, and other people also get the benefit of having a Home Manager module for that program

**Comment by trueNAHO** [comment]
> It may be more productive to contribute new options to Home Manager for the modules where we're currently using `home.file`, then we can use `lib.mkAfter` on them as we do for other modules, and other people also get the benefit of having a Home Manager module for that program

Yes. This has already been briefly discussed in some occasions, but there does not seem to be dedicated issue for improving integration with other platforms by upstreaming options.

AFAIK, this was first mentioned in https://github.com/danth/stylix/issues/159#issuecomment-1937879216.
